File: simulacion.py
import numpy as np
import random
import tqdm
import math
import matplotlib.pyplot as plt
from typing import Tuple


def numeros_enteros(max_entero: int = 1000) -> Tuple[int, int]:
  ''' Genera dos números enteros positivos aleatorios '''
  lista_numeros = range(1, max_entero+1)
  a = random.choice(lista_numeros)
  b = random.choice(lista_numeros)
  return a, b


def son_coprimos(a: int, b: int) -> bool:
  ''' Verifica si dos números, a y b, son coprimos '''
  mcd = math.gcd(a, b)
  return mcd == 1


def estima_probabilidad(max_entero: int = 1000, num_iteraciones: int = 10000) -> float:
  ''' estima la probabilidad de que dos números sean coprimos '''
  exitos = 0
  for _ in range(num_iteraciones):
    a, b = numeros_enteros(max_entero=max_entero)
    if son_coprimos(a=a, b=b):
      exitos += 1
  return exitos / num_iteraciones


iteraciones = range(1, 10000)
probabilidades = [estima_probabilidad(num_iteraciones=i) for i in tqdm.tqdm(iteraciones)]

plt.plot(iteraciones, probabilidades)
plt.axhline(y=6/np.pi**2, color='r', linestyle='-')

File: simulacion.R
library(FRACTION)


numeros_enteros <- function(max_entero = 1000){
  # Generamos una lista de posibles numeros enteros
  posibles_numeros <- sample(1:max_entero)

  # Extraemos aleatoriamente 2 elementos de esta lista
  a <- sample(posibles_numeros, 1, replace = TRUE)
  b <- sample(posibles_numeros, 1, replace = TRUE)
  return(list(a = a, b = b))
}


son_coprimos <- function(a, b){
  maximo_comun_divisor <- gcd(a, b)
  return(maximo_comun_divisor == 1)
}


estima_probabilidad <- function(max_entero = 1000, num_iteraciones = 1000){
  exito <- 0
  for(iter in seq(num_iteraciones)){
    numeros <- numeros_enteros(max_entero = max_entero)
    a <- numeros$a
    b <- numeros$b
    if(son_coprimos(a = a, b = b)){
      exito <- exito + 1
    }
  }
  return(exito / num_iteraciones)
}


iteraciones <- 1:1000
probabilidades <- c()

for(num_iter in iteraciones){
  prob <- estima_probabilidad(num_iteraciones = num_iter)
  probabilidades <- c(probabilidades, prob)
}


plot(probabilidades, type = 'l')
abline(h = 6/pi^2, col = 'red')
